File: java/ZigZagConversion.java
import java.util.stream.IntStream;

/**
 * https://leetcode.com/problems/zigzag-conversion/description/
 */

public class ZigZagConversion {
    public String convert(String s, int numRows) {
        if (numRows == 1 || s.isEmpty()) {
            return s;
        }
        String res = "";

        int diff = 2*(numRows - 1);
        int index = 0;

        //0th row
        while (index < s.length()) {
            res += s.charAt(index);
            index += diff;
        }

        int midDiff;
        //1 to numRows-2
        for (int i = 1; i < numRows-1; i++) {
            midDiff = diff - 2*i;
            index = i;
            while (index < s.length()) {
                res += s.charAt(index);
                if (index+midDiff < s.length()) {
                    res += s.charAt(index+midDiff);
                }
                index += diff;
            }
        }

        //numRows-1 th row
        index = numRows-1;
        while (index < s.length()) {
            res += s.charAt(index);
            index += diff;
        }
        return res;
    }
}

File: python3/submissions/zig_zag_conversion.py
# https://leetcode.com/problems/zigzag-conversion/

class Solution:
    def convert(self, s: str, numRows: int) -> str:
        if not s or len(s) < 2 or numRows == 1:
            return s
        res = ''
        length = len(s)
        adder = 2*numRows-2
        for i in range(numRows):
            extra_needed = i != 0 and i != numRows-1
            looper = i
            while looper < length or looper-2*i < length:
                if extra_needed and 0 < looper - 2 * i < length:
                    res = res + s[looper-2*i]
                if looper < length:
                    res = res + s[looper]
                looper = looper+adder
        return res
